[Spot] Better spot logs

Michaelvll opened the pull request:
This is to simplify the spot logs when waiting for the logs to be ready.

Tested:
- [x] `sky spot launch --num-nodes=2 'sleep 10000'`
Current:
<img width="1143" alt="image" src="https://user-images.githubusercontent.com/6753189/201551126-80034aa2-2201-4267-8af8-5e2b90b3105b.png">

<img width="1170" alt="image" src="https://user-images.githubusercontent.com/6753189/201551432-16836990-9ee8-430d-861d-d845065dd1a6.png">

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -232,56 +237,65 @@ def stream_logs_by_id(job_id: int, follow: bool = True) -> str:
+                logger.debug(
+                    f'INFO: The log is not ready yet{status_help_str}. '
+                    f'Waiting for {JOB_STATUS_CHECK_GAP_SECONDS} seconds.')
+                status_display.update('[bold cyan]Waiting for the spot job'
```
How about "Waiting for the spot job to start .."? Without "to start" it reads a bit unclear.

Michaelvll replied:
Good point! Fxied. Thanks!

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -232,56 +237,65 @@ def stream_logs_by_id(job_id: int, follow: bool = True) -> str:
+                status_help_str = ''
+                if (spot_status is not None and
+                        spot_status != spot_state.SpotStatus.RUNNING):
+                    status_help_str = f', status: {spot_status.value}'
```
```suggestion
                    status_help_str = f' (job status: {spot_status.value})'
```

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -199,22 +200,26 @@ def cancel_job_by_name(job_name: str) -> str:
-    while job_status is None:
-        logger.info('Waiting for the spot job to be started.')
-        time.sleep(1)
+    status_msg = ('[bold cyan]Waiting for controller, controller status: {}.[/]'
```
```suggestion
    status_msg = ('[bold cyan]Waiting for controller process to start/to be RUNNING (status: {}).[/]'
```
Reasons:
- distinguish "controller" and "controller process" 
  - We may want to avoid this confusion by renaming "controller process" to "monitor process" (for future)
- "waiting for X" is unclear to me

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -199,22 +200,26 @@ def cancel_job_by_name(job_name: str) -> str:
+            status_str = 'None'
+            if controller_status is not None:
+                status_str = controller_status.value
+            status_display.update(status_msg.update(status_str))
```
Q: why can we do `status_msg.update(status_str)`? `''.update()` says str doesn't have an update func?

Michaelvll replied:
Oops, that was a typo. Fixed. Thanks!

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -199,22 +200,26 @@ def cancel_job_by_name(job_name: str) -> str:
-        time.sleep(1)
+    status_msg = ('[bold cyan]Waiting for controller, controller status: {}.[/]'
+                  'It may take a few minutes.')
+    status_display = rich.status.Status('[bold cyan]Waiting for logs...[/]')
```
Q: I can't seem to trigger this log with `sky spot launch sleep 10000`.

Michaelvll replied:
It is the general initial status to avoid empty logging and will be soon updated later. I changed it to `Waiting for controller process to be RUNNING`, so it is consistent with the first logging afterwards.

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -199,22 +200,26 @@ def cancel_job_by_name(job_name: str) -> str:
+            time.sleep(_LOG_STREAM_CHECK_CONTROLLER_GAP_SECONDS)
+            controller_status = job_lib.get_status(job_id)
+
+        status_display.update(
+            '[bold cyan]Waiting for the spot job to be started[/]')
```
Q: I can't seem to trigger this log with `sky spot launch sleep 10000`.

Michaelvll replied:
That depends on when we run the sky spot logs. If we run the `sky spot logs 2` before the `sky spot launch sleep 1000`, this logs will appear. : )

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -232,56 +237,65 @@ def stream_logs_by_id(job_id: int, follow: bool = True) -> str:
+                logger.debug(
+                    f'INFO: (Log streaming) Got return code {returncode}. '
+                    f'Retrying in {JOB_STATUS_CHECK_GAP_SECONDS} seconds.')
+            status_display.update('[bold cyan]Waiting for the spot job[/]')
```
Q: I can't seem to trigger this log with sky spot launch sleep 10000.

Add "to start"?

Michaelvll replied:
This only happens when the cluster is preempted, i.e. the `tail_logs` returns nonzero.

concretevitamin reviewed (commented):
Thanks @Michaelvll - this helps the UX a bunch. Some questions.

(For the future: if more people start using spot_launch() , we may want to use the original logging?)

concretevitamin commented:
<img width="840" alt="Screen Shot 2022-11-18 at 18 16 42" src="https://user-images.githubusercontent.com/592670/202829675-f01c1d79-e488-46e5-a4d5-ac1dfcc09389.png">

I now see the `⠋ Waiting for the spot jobI 11-19...` line, which got printed during the log streaming of a `sleep 1000` job and after a `sky spot cancel -a` was issued.

The line joining can be fixed, but maybe we shouldn't show "Waiting for the spot job" in this case at all?

Michaelvll commented:
Thank you for the observations @concretevitamin! I have addressed the comments, but I will tag this PR blocked until #1433 is merged, since that may change the logging logic.

concretevitamin reviewed (approved):
LGTM, thanks @Michaelvll.

concretevitamin commented:
One minor thing:
```
» sky spot launch sleep 10000
...
I 11-27 03:16:03 log_lib.py:388] Start streaming logs for spot job 4.
INFO: Tip: use Ctrl-C to exit log streaming (task will not be killed).
INFO: Waiting for task resources on 1 node. This will block if the cluster is full.
INFO: All task resources reserved.
INFO: Reserved IPs: ['172.31.39.162']
```
Once the cluster is manually terminated on the console (simulating a preemption), the following lines immediately appear:
```
Shared connection to 3.134.104.105 closed.
⠦ Waiting for the job to start. It may take minutes.
```
Then the last line with the spinning cursor will disappear. It hangs for a few seconds, before showing:
```
ssh: connect to host 3.134.104.105 port 22: Connection timed out
⠹ Waiting for the job to start. It may take minutes.
```
the last line will somewhat quickly transition to
```
⠼ Waiting for the job to start (status: RECOVERING). It may take minutes.
```

The second snippet feels like new behavior? It's a bit weird to show it without either quickly transitioning to RECOVERING or showing "it may be preempted". Subjective, up to you.

Michaelvll commented:
Thank you for the review @concretevitamin! I could not reproduce the issue you mentioned, but likely, it is because the time we wait for the spot status update is not long enough. I increased it to `2.1 * JOB_STATUS_CHECK_GAP_SECONDS`, could you test it again to see if the problem mentioned above still happens?

concretevitamin commented:
I can still reproduce that snippet. Maybe we should debug live offline?

Michaelvll commented:
Updated the message to the following when preempted. PTAL : )
![image](https://user-images.githubusercontent.com/6753189/204709988-d84e4804-86bb-40c3-b6ad-9711e992dd3f.png)

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -286,15 +295,20 @@ def stream_logs_by_id(job_id: int, follow: bool = True) -> str:
-                _JOB_WAITING_STATUS_MESSAGE.format(status_str=''))
+            logger.info(
+                f'{colorama.Fore.YELLOW}The job is preempted or cancelled. '
+                'Waiting for its status to be updated.'
```
Maybe we can leave this part out? It's in the spinner line.

Michaelvll replied:
Good point! Removed it.

concretevitamin commented on sky/spot/spot_utils.py:
```diff
@@ -286,15 +295,20 @@ def stream_logs_by_id(job_id: int, follow: bool = True) -> str:
-            status_display.update(
-                _JOB_WAITING_STATUS_MESSAGE.format(status_str=''))
+            logger.info(
+                f'{colorama.Fore.YELLOW}The job is preempted or cancelled. '
```
I tried canceling a `sleep 1000` job and didn't see this message being printed. Saw it when manually terminating the spot VM.

Michaelvll replied:
Ahh, good point. Since we recently changed the behavior of `sky spot cancel`, this will only happen for preemption.

concretevitamin reviewed (approved):
LGTM, thanks @Michaelvll!
